src/terms/hubbard.jl

```diff
@@ -0,0 +1,178 @@
+using LinearAlgebra
+using Random
+
+"""
+Structure for manifold choice and projectors extraction.
+
+Overview of fields:
+- `iatom`: Atom position in the atoms array.
+- `species`: Chemical Element as in ElementPsp.
+- `label`: Orbital name, e.g.: "3S".
+
+All fields are optional, only the given ones will be used for selection.
+Can be called with an orbital NamedTuple and returns a boolean
+  stating whether the orbital belongs to the manifold.
+"""
+@kwdef struct OrbitalManifold
+    iatom   ::Union{Int64,  Nothing} = nothing
+    species ::Union{Symbol, AtomsBase.ChemicalSpecies, Nothing} = nothing
+    label   ::Union{String, Nothing} = nothing
+end
+function (s::OrbitalManifold)(orbital)
+    iatom_match    = isnothing(s.iatom)   || (s.iatom == orbital.iatom)
+    species_match  = isnothing(s.species) || (s.species == orbital.species)
+    label_match    = isnothing(s.label)   || (s.label == orbital.label)
+    iatom_match && species_match && label_match
+end
+
+function extract_manifold(basis::PlaneWaveBasis{T}, projectors, labels,
+                          manifold::OrbitalManifold) where {T}
+    # We extract the labels only for orbitals belonging to the manifold
+    manifold_labels = filter(manifold, labels)
+    isempty(manifold_labels) && @warn "Projector for $(manifold) not found."
+    proj_indices = findall(orb -> manifold(orb), labels)
+    manifold_projectors = map(enumerate(projectors)) do (ik, projk)
+        projk[:, proj_indices]
+    end
+    (; manifold_labels, manifold_projectors)
+end
+
+"""
+    compute_nhubbard(manifold, basis, ψ, occupation; [projectors, labels, positions])
+
+Computes a matrix nhubbard of size (n_spin, natoms, natoms), where each entry nhubbard[iatom, jatom]
+  contains the submatrix of the occupation matrix corresponding to the projectors
+  of atom iatom and atom jatom, with dimensions determined by the number of projectors for each atom.
+  The atoms and orbitals are defined by the manifold tuple.
+
+    nhubbard[σ, iatom, jatom][m1, m2] = Σₖ₍ₛₚᵢₙ₎Σₙ weights[ik, ibnd] * ψₙₖ' * Pᵢₘ₁ * Pᵢₘ₂' * ψₙₖ
+
+  where n or ibnd is the band index, ``weights[ik ibnd] = kweights[ik] * occupation[ik, ibnd]``
+  and ``Pᵢₘ₁`` is the pseudoatomic orbital projector for atom i and orbital m₁
+  (usually just the magnetic quantum number, since l is usually fixed).
```

Comment:
This is not a "usually" it's enforced in the code, right? So it's actually enforced that each Hubbard term only deals with one type of atomic orbital. This needs to be enforced by the design.